docs/DesignDocs/AttestedTLSAPIs.md

```diff
@@ -0,0 +1,758 @@
+
+Specification
+--------
+
+This document proposes the following new APIs that address the abovementioned limitations.
+- `oe_get_passport_attestation_certificate`
+- `oe_get_background_check_attestation_certificate`
+- `oe_parse_attestation_ceritiface`
+
+Note that the certificate generated by `oe_get_passport_attestation_certificate`
+and `oe_get_background_check_attestation_certificate`
+does not provide backward compatibility to existing verification APIs
+(e.g., `oe_verify_attestation_certificate_with_evidence_v2`
+and `oe_verify_attestation_certificate`).
+
+- oe_get_passport_attestation_certificate
+  ```c
+  /**
+   * oe_get_passport_attestation_certificate
+   *
+   * This function generates a self-signed X.509 certificate with embedded attestation results
+   * returned by a verifier. Note that the attestation results is an opaque buffer. The rule
+   * to interpret the buffer depends on the agreement between the relying party and the verifier.
+   *
+   * @experimental
+   *
+   * @param[in] format_id The format id of the evidence to be generated.
+   * @param[in] subject_name A string containing an X.509 distinguished
+   * name (DN) for customizing the generated certificate. This name is also used
+   * as the issuer name because this is a self-signed certificate.
+   * See RFC5280 (https://tools.ietf.org/html/rfc5280) for details.
+   * Example value "CN=Open Enclave SDK,O=OESDK TLS,C=US"
+   * @param[in] private_key A private key used to sign this certificate.
+   * @param[in] private_key_size The size of the buffer above.
+   * @param[in] public_key A public key used as the certificate's subject key.
+   * @param[in] public_key_size The size of the buffer above.
+   * @param[in] attestation_results_buffer An optional flat buffer representing the attestation
+   * results from a verifier.
+   * @param[in] attestation_results_buffer_size The size of the buffer above in bytes.
+   * @param[in] optional_parameters The optional format-specific input parameters.
+   * @param[in] optional_parameters_size The size of buffer above in bytes.
+   * @param[out] output_certificate A pointer to buffer pointer.
+   * @param[out] output_certificate_size The pointer to the size of the buffer above.
+   *
+   * @return OE_OK The operation was successful.
+   */
+  oe_result_t oe_get_passport_attestation_certificate(
+      const oe_uuid_t* format_id,
+      const unsigned char* subject_name,
+      const uint8_t* private_key,
+      size_t private_key_size,
+      const uint8_t* public_key,
+      size_t public_key_size,
+      const uint8_t* attestion_results_buffer,
+      size_t attestation_results_buffer_size,
+      uint8_t** output_certificate,
+      size_t* output_certificate_size);
+  ```
+  The API generates a certificate with an embedded attestation results that is specifically used
+  by an attested TLS with the passport model.
+  An important note is that the API does not enforce the binding between the
```

Comment:
Is there a reason for not enforcing the binding?

Reply:
In the passport model, the enforcement requires
- The attester enforces the binding into the evidence, sends to the verifier, and uses the returned attestation results to generate the certificate
- The relying party extracts the attestation results from the certificate, and validates the binding.

The above two steps require the attester and the relying party agree on a specific format of the attestation results (e.g., MAA otkens). OE currently does not have APIs support a specific verifier, which can surely be added in the future.

Reply:
In some cases the attester might need to "know" to ask the verifier for format in a particular format needed by a relying party (e.g., to ask for them in JWT vs CWT format), but it doesn't need to understand the format itself (in terms of ability to parse it) since it is an opaque buffer to the attester.